tests/unit/test_utils.py

```diff
@@ -940,5 +950,135 @@ def test_uses_original_endpoint_scheme(self):
             'http://s3-accelerate.amazonaws.com/foo/key.txt')
 
 
-if __name__ == '__main__':
-    unittest.main()
+class TestS3RegionRedirector(unittest.TestCase):
+    def setUp(self):
+        self.endpoint_bridge = mock.Mock()
+        self.endpoint_bridge.resolve.return_value = {
+            'endpoint_url': 'https://eu-central-1.amazonaws.com'
+        }
+        self.redirector = S3RegionRedirector(self.endpoint_bridge)
+        self.http_response = mock.Mock()
+
+    def test_set_request_url(self):
+        params = {'url': 'https://us-west-2.amazonaws.com/foo'}
+        context = {'signing': {
+            'endpoint': 'https://eu-central-1.amazonaws.com'
+        }}
+        self.redirector.set_request_url(params, context)
+        self.assertEqual(
+            params['url'], 'https://eu-central-1.amazonaws.com/foo')
+
+    def test_only_changes_request_url_if_endpoint_present(self):
+        params = {'url': 'https://us-west-2.amazonaws.com/foo'}
+        context = {}
+        self.redirector.set_request_url(params, context)
+        self.assertEqual(
+            params['url'], 'https://us-west-2.amazonaws.com/foo')
+
+    def test_sets_signing_context_from_cache(self):
+        signing_context = {'endpoint': 'bar'}
+        cache = {'foo': signing_context}
+        params = {'Bucket': 'foo'}
+        context = {}
+        self.redirector._cache = cache
```

Comment:
We should not be relying on modifying internal properties. Have you considered being able to inject a cache or make the property public?